Source/Core/Core/State.cpp

```diff
@@ -578,62 +745,50 @@ static void LoadFileStateData(const std::string& filename, std::vector<u8>& ret_
   }
 
   StateHeader header;
-  if (!f.ReadArray(&header, 1))
-  {
-    Core::DisplayMessage("State not found", 2000);
+  if (!ReadStateHeaderFromFile(header, f) || !ValidateHeaders(header))
     return;
-  }
 
-  if (strncmp(SConfig::GetInstance().GetGameID().c_str(), header.gameID, 6))
+  StateExtendedHeader extended_header;
+  f.ReadArray(&extended_header.base_header, 1);
+  // If StateExtendedHeader is ammended to include more than the base, add ReadBytes() calls here.
+  // Do NOT use ReadArray, as this would prevent future revisions from reading in char arrays.
+
+  if (extended_header.base_header.header_version != EXTENDED_HEADER_VERSION)
   {
-    Core::DisplayMessage(fmt::format("State belongs to a different game (ID {})",
-                                     std::string_view{header.gameID, std::size(header.gameID)}),
-                         2000);
-    return;
+    const std::string message =
+        fmt::format("Savestate extended header mismatch!\n\n"
+                    "Current: {}\n"
+                    "Savestate: {}",
+                    EXTENDED_HEADER_VERSION, extended_header.base_header.header_version);
+    Core::DisplayMessage(message, OSD::Duration::NORMAL);
   }
 
   std::vector<u8> buffer;
 
-  if (header.size != 0)  // non-zero size means the state is compressed
+  switch (extended_header.base_header.compression_type)
+  {
+  case (CompressionType::LZ4):
   {
     Core::DisplayMessage("Decompressing State...", 500);
-
-    buffer.resize(header.size);
-
-    lzo_uint i = 0;
-    while (true)
-    {
-      lzo_uint32 cur_len = 0;  // number of bytes to read
-      lzo_uint new_len = 0;    // number of bytes to write
-
-      if (!f.ReadArray(&cur_len, 1))
-        break;
-
-      f.ReadBytes(out, cur_len);
-      const int res = lzo1x_decompress(out, cur_len, &buffer[i], &new_len, nullptr);
-      if (res != LZO_E_OK)
-      {
-        // This doesn't seem to happen anymore.
-        PanicAlertFmtT("Internal LZO Error - decompression failed ({0}) ({1}, {2}) \n"
-                       "Try loading the state again",
-                       res, i, new_len);
-        return;
-      }
-
-      i += new_len;
-    }
+    DecompressLZ4(buffer, extended_header.base_header.uncompressed_size, f);
```

Comment:
`DecompressLZ4()` should return whether it succeeded or not, you don't want partially/incorrectly compressed data to be returned.